iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/relational/planner/distribute/ExchangeNodeGenerator.java

```diff
@@ -45,7 +47,37 @@ public class ExchangeNodeGenerator extends SimplePlanRewriter<ExchangeNodeGenera
   @Override
   public List<PlanNode> visitTableScan(TableScanNode node, PlanContext context) {
 
-    if (node.getRegionReplicaSetList().size() > 1) {
+    Map<TRegionReplicaSet, TableScanNode> tableScanNodeMap = new HashMap<>();
+
+    for (DeviceEntry deviceEntry : node.getDeviceEntries()) {
+      List<TRegionReplicaSet> regionReplicaSets =
+          context
+              .analysis
+              .getDataPartitionInfo()
+              .getDataRegionReplicaSetWithTimeFilter(
+                  deviceEntry.getDeviceID(), node.getTimeFilter());
+      for (TRegionReplicaSet regionReplicaSet : regionReplicaSets) {
+        TableScanNode tableScanNode =
+            tableScanNodeMap.computeIfAbsent(
+                regionReplicaSet,
+                k ->
+                    new TableScanNode(
+                        context.queryContext.getQueryId().genPlanNodeId(),
+                        node.getQualifiedObjectName(),
+                        node.getOutputSymbols(),
+                        node.getAssignments(),
+                        new ArrayList<>(),
+                        node.getIdAndAttributeIndexMap(),
+                        node.getScanOrder(),
+                        node.getTimePredicate().orElse(null),
+                        node.getPushDownPredicate()));
+        tableScanNode.appendDeviceEntry(deviceEntry);
+        tableScanNode.setRegionReplicaSet(regionReplicaSet);
```

Comment:
no need to set each time for one device, you can put the set logic in `computeIfAbsent`